File: python/callisto_sdk/resources/otp.py
from __future__ import annotations

from typing import Any, Optional, Union

from .._http import Transport
from ..errors import ValidationError
from ..models import Otp, OtpProvider, OtpType, Paginated, SendOtpResult, VerifyOtpResult


class OtpResource:
    def __init__(self, transport: Transport):
        self._t = transport

    def send(
        self,
        to: str,
        message: str,
        sender: Optional[str] = None,
        expired_in: Optional[int] = None,
        type: Optional[Union[OtpType, str]] = None,
        digit_size: Optional[int] = None,
        provider: Optional[Union[OtpProvider, str]] = None,
        instance_code: Optional[str] = None,
    ) -> SendOtpResult:
        provider_value = provider.value if isinstance(provider, OtpProvider) else provider
        if provider_value == OtpProvider.WHATSAPP.value and not instance_code:
            raise ValidationError("instance_code is required when provider is whatsapp")
        body: dict[str, Any] = {"to": to, "message": message}
        if sender is not None:
            body["sender"] = sender
        if expired_in is not None:
            body["expired_in"] = expired_in
        if type is not None:
            body["type"] = type.value if isinstance(type, OtpType) else type
        if digit_size is not None:
            body["digit_size"] = digit_size
        if provider_value is not None:
            body["provider"] = provider_value
        if instance_code is not None:
            body["instanceCode"] = instance_code
        return SendOtpResult.from_dict(self._t.request("POST", "/otp/send", body=body))

    def verify(self, otp_id: str, code: str) -> VerifyOtpResult:
        return VerifyOtpResult.from_dict(
            self._t.request("POST", "/otp/verify", body={"otp_id": otp_id, "code": code})
        )

    def get_status(self, otp_id: str) -> Otp:
        return Otp.from_dict(self._t.request("GET", f"/otps/{otp_id}"))

    def list(
        self,
        started_at: Optional[str] = None,
        ended_at: Optional[str] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> Paginated[Otp]:
        return Paginated.from_dict(self._t.request("GET", "/otps", query={
            "started_at": started_at, "ended_at": ended_at, "page": page, "limit": limit,
        }), Otp.from_dict)

File: java/src/main/java/com/callisto/sdk/resources/Otp.java
package com.callisto.sdk.resources;

import com.callisto.sdk.enums.OtpProvider;
import com.callisto.sdk.errors.ValidationException;
import com.callisto.sdk.http.Transport;
import com.callisto.sdk.models.Paginated;
import com.callisto.sdk.models.SendOtpResult;
import com.callisto.sdk.models.VerifyOtpResult;

import java.util.LinkedHashMap;
import java.util.Map;

/** The OTP resource. */
public class Otp {

    private final Transport transport;

    public Otp(Transport transport) {
        this.transport = transport;
    }

    /**
     * Generates and sends a one-time password.
     *
     * <p>When {@code provider == whatsapp}, {@code instanceCode} is required; otherwise a
     * {@link ValidationException} is thrown before any request is made. {@code instanceCode}
     * is sent to the API as {@code instanceCode}.
     *
     * @param request send options.
     */
    public SendOtpResult send(OtpSendRequest request) {
        String provider = request.getProvider();
        if (OtpProvider.WHATSAPP.getValue().equals(provider)
                && (request.getInstanceCode() == null || request.getInstanceCode().isEmpty())) {
            ValidationException ex =
                    new ValidationException("instanceCode is required when provider is whatsapp");
            if (transport.reporter() != null) {
                transport.reporter().captureException(ex);
            }
            throw ex;
        }
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("to", request.getTo());
        body.put("message", request.getMessage());
        if (request.getSender() != null) {
            body.put("sender", request.getSender());
        }
        if (request.getExpiredIn() != null) {
            body.put("expired_in", request.getExpiredIn());
        }
        if (request.getType() != null) {
            body.put("type", request.getType());
        }
        if (request.getDigitSize() != null) {
            body.put("digit_size", request.getDigitSize());
        }
        if (provider != null) {
            body.put("provider", provider);
        }
        if (request.getInstanceCode() != null) {
            body.put("instanceCode", request.getInstanceCode());
        }
        return transport.mapper().convertValue(
                transport.request("POST", "/otp/send", body, null), SendOtpResult.class);
    }

    /**
     * Verifies a code against an OTP.
     *
     * @param otpId the OTP ID returned by {@code send}.
     * @param code  the code submitted by the user.
     */
    public VerifyOtpResult verify(String otpId, String code) {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("otp_id", otpId);
        body.put("code", code);
        return transport.mapper().convertValue(
                transport.request("POST", "/otp/verify", body, null), VerifyOtpResult.class);
    }

    /**
     * Fetches a single OTP by ID.
     *
     * @param otpId the OTP ID.
     */
    public com.callisto.sdk.models.Otp getStatus(String otpId) {
        return transport.mapper().convertValue(
                transport.request("GET", "/otps/" + otpId, null, null),
                com.callisto.sdk.models.Otp.class);
    }

    /** Lists OTPs with no filters. */
    public Paginated<com.callisto.sdk.models.Otp> list() {
        return list(null, null, null, null);
    }

    /**
     * Lists OTPs.
     *
     * @param startedAt optional start date/time filter.
     * @param endedAt   optional end date/time filter.
     * @param page      optional page number.
     * @param limit     optional items per page (sent as query key {@code limit}).
     */
    public Paginated<com.callisto.sdk.models.Otp> list(String startedAt, String endedAt,
                                                       Integer page, Integer limit) {
        Map<String, Object> query = new LinkedHashMap<>();
        query.put("started_at", startedAt);
        query.put("ended_at", endedAt);
        query.put("page", page);
        query.put("limit", limit);
        return Paginated.fromJson(
                transport.request("GET", "/otps", null, query),
                transport.mapper(), com.callisto.sdk.models.Otp.class);
    }
}
